// @textlint/kernel/module/textlint-kernel.js
// MIT © 2017- azu
import * as assert from "assert";
// sequence
import FixerProcessor from "./fixer/fixer-processor";
// parallel
import LinterProcessor from "./linter/linter-processor";
// message process manager
import MessageProcessManager from "./messages/MessageProcessManager";
import filterIgnoredProcess from "./messages/filter-ignored-process";
import filterDuplicatedProcess from "./messages/filter-duplicated-process";
import filterSeverityProcess from "./messages/filter-severity-process";
import sortMessageProcess from "./messages/sort-messages-process";
import { TextlintKernelDescriptor } from "./descriptor";
import { TextlintSourceCodeImpl } from "./context/TextlintSourceCodeImpl";
import { isTxtAST } from "@textlint/ast-tester";
import _debug from "debug";
var debug = _debug("textlint:kernel");
/**
 * add fileName to trailing of error message
 * @param {string|undefined} fileName
 * @param {string} message
 * @returns {string}
 */
function addingAtFileNameToError(fileName, message) {
    if (!fileName) {
        return message;
    }
    return message + "\nat " + fileName;
}
/**
 * TextlintKernel is core logic written by pure JavaScript.
 *
 * Pass
 *
 * - config
 * - plugins
 * - rules
 * - filterRules
 * - messageProcessor
 *
 */
var TextlintKernel = /** @class */ (function () {
    /**
     * @param config
     */
    function TextlintKernel(config) {
        if (config === void 0) { config = {}; }
        // this.config often is undefined.
        this.config = config;
        // Initialize Message Processor
        // Now, It it built-in process only
        // filter `shouldIgnore()` results
        this.messageProcessManager = new MessageProcessManager([filterIgnoredProcess]);
        // filter duplicated messages
        this.messageProcessManager.add(filterDuplicatedProcess);
        // filter by severity
        this.messageProcessManager.add(filterSeverityProcess(this.config));
        this.messageProcessManager.add(sortMessageProcess);
    }
    /**
     * lint text by registered rules.
     * The result contains target filePath and error messages.
     * @param {string} text
     * @param {Object} options linting options
     * @returns {Promise.<TextlintResult>}
     */
    TextlintKernel.prototype.lintText = function (text, options) {
        var _this = this;
        return Promise.resolve().then(function () {
            var descriptor = new TextlintKernelDescriptor({
                rules: options.rules || [],
                filterRules: options.filterRules || [],
                plugins: options.plugins || []
            });
            return _this._parallelProcess({
                descriptor: descriptor,
                text: text,
                options: options
            });
        });
    };
    /**
     * fix texts and return fix result object
     * @param {string} text
     * @param {Object} options lint options
     * @returns {Promise.<TextlintFixResult>}
     */
    TextlintKernel.prototype.fixText = function (text, options) {
        var _this = this;
        return Promise.resolve().then(function () {
            var descriptor = new TextlintKernelDescriptor({
                rules: options.rules || [],
                filterRules: options.filterRules || [],
                plugins: options.plugins || []
            });
            return _this._sequenceProcess({
                descriptor: descriptor,
                options: options,
                text: text
            });
        });
    };
    /**
     * process text in parallel for Rules and return {Promise.<TextLintResult>}
     * In other word, parallel flow process.
     * @param {*} processor
     * @param {string} text
     * @param {Object} options
     * @returns {Promise.<TextlintResult>}
     * @private
     */
    TextlintKernel.prototype._parallelProcess = function (_a) {
        var descriptor = _a.descriptor, text = _a.text, options = _a.options;
        var ext = options.ext, filePath = options.filePath, configBaseDir = options.configBaseDir;
        var plugin = descriptor.findPluginDescriptorWithExt(ext);
        debug("available extensions: %o", descriptor.availableExtensions);
        if (plugin === undefined) {
            throw new Error("Not found available plugin for " + ext);
        }
        debug("use plugin: %s", plugin.id);
        var processor = plugin.processor;
        var _b = processor.processor(ext), preProcess = _b.preProcess, postProcess = _b.postProcess;
        assert.ok(typeof preProcess === "function" && typeof postProcess === "function", "processor should implements {preProcess, postProcess}");
        var preProcessResult = preProcess(text, filePath);
        var isPluginReturnAnAST = isTxtAST(preProcessResult);
        var textForAST = isPluginReturnAnAST ? text : preProcessResult.text;
        var ast = isPluginReturnAnAST ? preProcessResult : preProcessResult.ast;
        var sourceCode = new TextlintSourceCodeImpl({
            text: textForAST,
            ast: ast,
            ext: ext,
            filePath: filePath
        });
        var linterProcessor = new LinterProcessor(processor, this.messageProcessManager);
        return linterProcessor
            .process({
            config: this.config,
            ruleDescriptors: descriptor.rule,
            filterRuleDescriptors: descriptor.filterRule,
            sourceCode: sourceCode,
            configBaseDir: configBaseDir
        })
            .catch(function (error) {
            error.message = addingAtFileNameToError(filePath, error.message);
            return Promise.reject(error);
        });
    };
    /**
     * process text in series for Rules and return {Promise.<TextlintFixResult>}
     * In other word, sequence flow process.
     * @param {*} processor
     * @param {string} text
     * @param {TextlintKernelOptions} options
     * @returns {Promise.<TextlintFixResult>}
     * @private
     */
    TextlintKernel.prototype._sequenceProcess = function (_a) {
        var descriptor = _a.descriptor, text = _a.text, options = _a.options;
        var ext = options.ext, filePath = options.filePath, configBaseDir = options.configBaseDir;
        var plugin = descriptor.findPluginDescriptorWithExt(ext);
        debug("available extensions: %o", descriptor.availableExtensions);
        if (plugin === undefined) {
            throw new Error("Not found available plugin for " + ext);
        }
        debug("use plugin: %s", plugin.id);
        var processor = plugin.processor;
        var _b = processor.processor(ext), preProcess = _b.preProcess, postProcess = _b.postProcess;
        assert.ok(typeof preProcess === "function" && typeof postProcess === "function", "processor should implements {preProcess, postProcess}");
        var preProcessResult = preProcess(text, filePath);
        var isPluginReturnAnAST = isTxtAST(preProcessResult);
        var textForAST = isPluginReturnAnAST ? text : preProcessResult.text;
        var ast = isPluginReturnAnAST ? preProcessResult : preProcessResult.ast;
        var sourceCode = new TextlintSourceCodeImpl({
            text: textForAST,
            ast: ast,
            ext: ext,
            filePath: filePath
        });
        var fixerProcessor = new FixerProcessor(processor, this.messageProcessManager);
        return fixerProcessor
            .process({
            config: this.config,
            ruleDescriptors: descriptor.rule,
            filterRules: descriptor.filterRule,
            sourceCode: sourceCode,
            configBaseDir: configBaseDir
        })
            .catch(function (error) {
            error.message = addingAtFileNameToError(filePath, error.message);
            return Promise.reject(error);
        });
    };
    return TextlintKernel;
}());
export { TextlintKernel };
