// textlint/lib/config/config-loader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadConfig = void 0;
var module_interop_1 = require("@textlint/module-interop");
var rc_config_loader_1 = require("rc-config-loader");
/**
 * @param configFileName "textlint" for .textlinrc
 * @param cwd current working dir
 * @param configFilePath it is preferred than configFileName
 * @param moduleResolver
 */
function loadConfig(_a) {
    var cwd = _a.cwd, configFileName = _a.configFileName, configFilePath = _a.configFilePath, moduleResolver = _a.moduleResolver;
    // if specify Config module, use it
    if (configFilePath) {
        try {
            var modulePath = moduleResolver.resolveConfigPackageName(configFilePath);
            return {
                config: module_interop_1.moduleInterop(require(modulePath)),
                filePath: modulePath
            };
        }
        catch (error) {
            // not found config module
        }
    }
    // auto or specify path to config file
    var result = rc_config_loader_1.rcFile(configFileName, {
        configFileName: configFilePath,
        defaultExtension: [".json", ".js", ".yml"],
        packageJSON: {
            fieldName: "textlint"
        },
        cwd: cwd
    });
    if (result === undefined) {
        return {
            config: {},
            filePath: undefined
        };
    }
    return {
        config: result.config,
        filePath: result.filePath
    };
}
exports.loadConfig = loadConfig;
