// @textlint/feature-flag/lib/index.js
// MIT © 2017 azu
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.throwIfTesting = exports.throwWithoutExperimental = exports.coreFlags = exports.isFeatureEnabled = exports.setFeature = exports.resetFlags = void 0;
var map_like_1 = require("map-like");
var flagMap = new map_like_1.MapLike();
/**
 * IT IS FOR TESTING
 */
var resetFlags = function () {
    flagMap.clear();
};
exports.resetFlags = resetFlags;
/**
 * set feature flag
 * @param {string} flagName
 * @param {boolean} status
 */
var setFeature = function (flagName, status) {
    flagMap.set(flagName, status);
};
exports.setFeature = setFeature;
/**
 * If the feature flag of `flagName` is enabled, return true.
 * @param {string} flagName
 * @param {boolean=false} [loose]
 * @returns {boolean}
 */
var isFeatureEnabled = function (flagName, _a) {
    var _b = _a === void 0 ? {} : _a, _c = _b.loose, loose = _c === void 0 ? false : _c;
    if (!flagMap.has(flagName)) {
        if (loose) {
            // loose-mode, return false
            return false;
        }
        else {
            throw new Error("@textlint/feature-flag:Error: " + flagName + " is not defined.");
        }
    }
    var status = flagMap.get(flagName);
    if (typeof status !== "boolean") {
        throw new Error("flag should be boolean, but it is :" + status);
    }
    return status;
};
exports.isFeatureEnabled = isFeatureEnabled;
// == CORE Flags
// if run textlint --experimental, set experimental true by default
if (typeof process !== "undefined" && Array.isArray(process.argv) && process.argv.indexOf("--experimental") !== -1) {
    exports.setFeature("core.experimental", true);
}
/**
 * Core flags
 * @type {{experimental, experimental, runningCLI, runningCLI, runningTester, runningTester}}
 */
exports.coreFlags = {
    // Experimental
    get experimental() {
        return exports.isFeatureEnabled("core.experimental", {
            loose: true
        });
    },
    set experimental(status) {
        exports.setFeature("core.experimental", status);
    },
    // CLI
    get runningCLI() {
        return exports.isFeatureEnabled("core.runningCLI", {
            loose: true
        });
    },
    set runningCLI(status) {
        exports.setFeature("core.runningCLI", status);
    },
    // textlint-tester
    get runningTester() {
        return exports.isFeatureEnabled("core.runningTester", {
            loose: true
        });
    },
    set runningTester(status) {
        exports.setFeature("core.runningTester", status);
    }
};
/**
 * if current is not experimental, throw error message.
 * @param message
 */
function throwWithoutExperimental(message) {
    if (exports.coreFlags.runningCLI && !exports.coreFlags.experimental) {
        throw Error(message);
    }
}
exports.throwWithoutExperimental = throwWithoutExperimental;
/**
 * if current is in testing, throw error message.
 * @param {string} message
 */
function throwIfTesting(message) {
    if (exports.coreFlags.runningTester) {
        throw Error(message);
    }
}
exports.throwIfTesting = throwIfTesting;
