// textlint/module/util/object-to-kernel-format.js
/**
 * Convert rulesObject to TextlintKernelRule
 * {
 *     "rule-name": rule
 * },
 * {
 *     "rule-name": ruleOption
 * }
 *
 * => TextlintKernelRule
 */
export var rulesObjectToKernelRule = function (rules, rulesOption) {
    return Object.keys(rules).map(function (ruleId) {
        return {
            ruleId: ruleId,
            rule: rules[ruleId],
            options: rulesOption[ruleId]
        };
    });
};
export var filterRulesObjectToKernelRule = function (rules, rulesOption) {
    return Object.keys(rules).map(function (ruleId) {
        return {
            ruleId: ruleId,
            rule: rules[ruleId],
            options: rulesOption[ruleId]
        };
    });
};
/**
 * Convert pluginsObject to TextlintKernelPlugin
 * {
 *     "plugin-name": plugin
 * },
 * {
 *     "plugin-name": pluginOption
 * }
 *
 * => TextlintKernelPlugin
 */
export var pluginsObjectToKernelRule = function (plugins, pluginsOption) {
    return Object.keys(plugins).map(function (pluginId) {
        return {
            pluginId: pluginId,
            plugin: plugins[pluginId],
            options: pluginsOption[pluginId]
        };
    });
};
