// @textlint/kernel/module/task/textlint-core-task.js
// LICENSE : MIT
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { TextlintRuleErrorImpl } from "../context/TextlintRuleErrorImpl";
import { PromiseEventEmitter } from "./promise-event-emitter";
import SourceLocation from "../core/source-location";
import timing from "../util/timing";
import MessageType from "../shared/type/MessageType";
import { EventEmitter } from "events";
import * as assert from "assert";
import { normalizeTextlintKeyPath } from "@textlint/utils";
import { TextlintRuleContextImpl } from "../context/TextlintRuleContextImpl";
import _debug from "debug";
import { Controller as TraverseController } from "@textlint/ast-traverse";
var traverseController = new TraverseController();
var debug = _debug("textlint:core-task");
var RuleTypeEmitter = /** @class */ (function (_super) {
    __extends(RuleTypeEmitter, _super);
    function RuleTypeEmitter() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return RuleTypeEmitter;
}(PromiseEventEmitter));
/**
 * CoreTask receive AST and prepare, traverse AST, emit nodeType event!
 * You can observe task and receive "message" event that is TextLintMessage.
 */
var TextLintCoreTask = /** @class */ (function (_super) {
    __extends(TextLintCoreTask, _super);
    function TextLintCoreTask() {
        var _this = _super.call(this) || this;
        _this.ruleTypeEmitter = new RuleTypeEmitter();
        return _this;
    }
    Object.defineProperty(TextLintCoreTask, "events", {
        get: function () {
            return {
                // receive start event
                start: "start",
                // receive message from each rules
                message: "message",
                // receive complete event
                complete: "complete",
                // receive error event
                error: "error"
            };
        },
        enumerable: false,
        configurable: true
    });
    TextLintCoreTask.prototype.createShouldIgnore = function () {
        var _this = this;
        var shouldIgnore = function (args) {
            var ruleId = args.ruleId, range = args.range, optional = args.optional;
            assert.ok(typeof range[0] !== "undefined" && typeof range[1] !== "undefined" && range[0] >= 0 && range[1] >= 0, "ignoreRange should have actual range: " + range);
            // FIXME: should have index, loc
            // should be compatible with LintReportedMessage?
            var message = {
                type: MessageType.ignore,
                ruleId: ruleId,
                range: range,
                // ignoring target ruleId - default: filter all messages
                // This ruleId should be normalized, because the user can report any value
                ignoringRuleId: optional.ruleId ? normalizeTextlintKeyPath(optional.ruleId) : "*"
            };
            _this.emit(TextLintCoreTask.events.message, message);
        };
        return shouldIgnore;
    };
    TextLintCoreTask.prototype.createReporter = function (sourceCode) {
        var _this = this;
        var sourceLocation = new SourceLocation(sourceCode);
        /**
         * push new RuleError to results
         * @param {ReportMessage} reportArgs
         */
        var reportFunction = function (reportArgs) {
            var ruleId = reportArgs.ruleId, severity = reportArgs.severity, ruleError = reportArgs.ruleError;
            debug("%s pushReport %s", ruleId, ruleError);
            var _a = sourceLocation.adjust(reportArgs), line = _a.line, column = _a.column, fix = _a.fix;
            var index = sourceCode.positionToIndex({ line: line, column: column });
            // add TextLintMessage
            var message = {
                type: MessageType.lint,
                ruleId: ruleId,
                message: ruleError.message,
                index: index,
                // See https://github.com/textlint/textlint/blob/master/typing/textlint.d.ts
                line: line,
                column: column + 1,
                severity: severity,
                fix: fix !== undefined ? fix : undefined
            };
            if (!(ruleError instanceof TextlintRuleErrorImpl)) {
                // FIXME: RuleReportedObject should be removed
                // `error` is a any data.
                var data = ruleError;
                message.data = data;
            }
            _this.emit(TextLintCoreTask.events.message, message);
        };
        return reportFunction;
    };
    /**
     * start process and emitting events.
     * You can listen message by `task.on("message", message => {})`
     * @param {SourceCode} sourceCode
     */
    TextLintCoreTask.prototype.startTraverser = function (sourceCode) {
        var _this = this;
        this.emit(TextLintCoreTask.events.start);
        var promiseQueue = [];
        var ruleTypeEmitter = this.ruleTypeEmitter;
        traverseController.traverse(sourceCode.ast, {
            enter: function (node, parent) {
                var type = node.type;
                Object.defineProperty(node, "parent", { value: parent });
                if (ruleTypeEmitter.listenerCount(type) > 0) {
                    var promise = ruleTypeEmitter.emit(type, node);
                    promiseQueue.push(promise);
                }
            },
            leave: function (node) {
                var type = node.type + ":exit";
                if (ruleTypeEmitter.listenerCount(type) > 0) {
                    var promise = ruleTypeEmitter.emit(type, node);
                    promiseQueue.push(promise);
                }
            }
        });
        Promise.all(promiseQueue)
            .then(function () {
            _this.emit(TextLintCoreTask.events.complete);
        })
            .catch(function (error) {
            _this.emit(TextLintCoreTask.events.error, error);
        });
    };
    /**
     * try to get rule object
     */
    TextLintCoreTask.prototype.tryToGetRuleObject = function (ruleCreator, ruleContext, ruleOptions) {
        try {
            return ruleCreator(ruleContext, ruleOptions);
        }
        catch (error) {
            error.message = "Error while loading rule '" + ruleContext.id + "': " + error.message;
            throw error;
        }
    };
    /**
     * try to get filter rule object
     */
    TextLintCoreTask.prototype.tryToGetFilterRuleObject = function (ruleCreator, ruleContext, ruleOptions) {
        try {
            return ruleCreator(ruleContext, ruleOptions);
        }
        catch (error) {
            error.message = "Error while loading filter rule '" + ruleContext.id + "': " + error.message;
            throw error;
        }
    };
    /**
     * add all the node types as listeners of the rule
     * @param {Function} ruleCreator
     * @param {Readonly<RuleContext>|Readonly<FilterRuleContext>} ruleContext
     * @param {Object|boolean|undefined} ruleOptions
     * @returns {Object}
     */
    TextLintCoreTask.prototype.tryToAddListenRule = function (ruleCreator, ruleContext, ruleOptions) {
        var _this = this;
        var ruleObject = ruleContext instanceof TextlintRuleContextImpl
            ? this.tryToGetRuleObject(ruleCreator, ruleContext, ruleOptions)
            : this.tryToGetFilterRuleObject(ruleCreator, ruleContext, ruleOptions);
        var types = Object.keys(ruleObject);
        types.forEach(function (nodeType) {
            _this.ruleTypeEmitter.on(nodeType, timing.enabled ? timing.time(ruleContext.id, ruleObject[nodeType]) : ruleObject[nodeType]);
        });
    };
    return TextLintCoreTask;
}(EventEmitter));
export default TextLintCoreTask;
