// textlint/lib/util/object-to-kernel-format.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pluginsObjectToKernelRule = exports.filterRulesObjectToKernelRule = exports.rulesObjectToKernelRule = void 0;
/**
 * Convert rulesObject to TextlintKernelRule
 * {
 *     "rule-name": rule
 * },
 * {
 *     "rule-name": ruleOption
 * }
 *
 * => TextlintKernelRule
 */
exports.rulesObjectToKernelRule = function (rules, rulesOption) {
    return Object.keys(rules).map(function (ruleId) {
        return {
            ruleId: ruleId,
            rule: rules[ruleId],
            options: rulesOption[ruleId]
        };
    });
};
exports.filterRulesObjectToKernelRule = function (rules, rulesOption) {
    return Object.keys(rules).map(function (ruleId) {
        return {
            ruleId: ruleId,
            rule: rules[ruleId],
            options: rulesOption[ruleId]
        };
    });
};
/**
 * Convert pluginsObject to TextlintKernelPlugin
 * {
 *     "plugin-name": plugin
 * },
 * {
 *     "plugin-name": pluginOption
 * }
 *
 * => TextlintKernelPlugin
 */
exports.pluginsObjectToKernelRule = function (plugins, pluginsOption) {
    return Object.keys(plugins).map(function (pluginId) {
        return {
            pluginId: pluginId,
            plugin: plugins[pluginId],
            options: pluginsOption[pluginId]
        };
    });
};
